pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -0,0 +1,253 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cm
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/sets"
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha1"
+	"k8s.io/kubernetes/pkg/kubelet/deviceplugin"
+)
+
+// podDevices represents a list of pod to Device mappings.
+type containerToDevice map[string]sets.String
+type podDevices struct {
+	podDeviceMapping map[string]containerToDevice
+}
+
+func newPodDevices() *podDevices {
+	return &podDevices{
+		podDeviceMapping: make(map[string]containerToDevice),
+	}
+}
+func (pdev *podDevices) pods() sets.String {
+	ret := sets.NewString()
+	for k := range pdev.podDeviceMapping {
+		ret.Insert(k)
+	}
+	return ret
+}
+
+func (pdev *podDevices) insert(podUID, contName string, device string) {
+	if _, exists := pdev.podDeviceMapping[podUID]; !exists {
+		pdev.podDeviceMapping[podUID] = make(containerToDevice)
+	}
+	if _, exists := pdev.podDeviceMapping[podUID][contName]; !exists {
+		pdev.podDeviceMapping[podUID][contName] = sets.NewString()
+	}
+	pdev.podDeviceMapping[podUID][contName].Insert(device)
+}
+
+func (pdev *podDevices) getDevices(podUID, contName string) sets.String {
+	containers, exists := pdev.podDeviceMapping[podUID]
+	if !exists {
+		return nil
+	}
+	devices, exists := containers[contName]
+	if !exists {
+		return nil
+	}
+	return devices
+}
+
+func (pdev *podDevices) delete(pods []string) {
+	for _, uid := range pods {
+		delete(pdev.podDeviceMapping, uid)
+	}
+}
+
+func (pdev *podDevices) devices() sets.String {
+	ret := sets.NewString()
+	for _, containerToDevice := range pdev.podDeviceMapping {
+		for _, deviceSet := range containerToDevice {
+			ret = ret.Union(deviceSet)
+		}
+	}
+	return ret
+}
+
+type DevicePluginHandler interface {
+	Start() error
+	Devices() map[string][]*pluginapi.Device
+	Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) ([]*pluginapi.AllocateResponse, error)
+}
+
+// updateCapacityCallback is called to update ContainerManager capacity when
+// device capacity changes.
+type updateCapacityCallback func(v1.ResourceList)
+
+type DevicePluginHandlerImpl struct {
+	sync.Mutex
+	devicePluginManager deviceplugin.Manager
+	// devicePluginManagerMonitorCallback is used for testing only.
+	devicePluginManagerMonitorCallback deviceplugin.MonitorCallback
+	// allDevices and allocated are keyed by resource_name.
+	allDevices         map[string]sets.String
+	allocatedDevices   map[string]*podDevices
+	updateCapacityFunc updateCapacityCallback
```

Comment:
Why does `updateCapacityFunc` need to be part of the struct? We can also get rid of the `updateCapacityCallback` type.